File: backend/app/schemas/accounting.py
# backend/app/schemas/accounting.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import date, datetime
from enum import Enum


class CategorieDepenseEnum(str, Enum):
    ACHAT_ANIMAUX = "achat_animaux"
    ACHAT_OEUFS = "achat_oeufs"
    ALIMENTATION = "alimentation"
    VACCINS_SOINS = "vaccins_soins"
    EQUIPEMENT = "equipement"
    PERSONNEL = "personnel"
    EAU_ELECTRICITE = "eau_electricite"
    ENTRETIEN = "entretien"
    COMPOSTAGE = "compostage"
    TRANSPORT = "transport"
    FRAIS_DIVERS = "frais_divers"


class CategorieRecetteEnum(str, Enum):
    VENTE_ANIMAUX_VIVANTS = "vente_animaux_vivants"
    VENTE_VIANDE = "vente_viande"
    VENTE_LAIT = "vente_lait"
    VENTE_LAINE = "vente_laine"
    VENTE_OEUFS = "vente_oeufs"
    VENTE_LARVES = "vente_larves"
    VENTE_COMPOST = "vente_compost"
    VENTE_FUMIER = "vente_fumier"
    SUBVENTIONS = "subventions"
    AUTRES = "autres"


class DepenseBase(BaseModel):
    categorie: CategorieDepenseEnum
    montant: float = Field(..., gt=0)
    date: date
    description: Optional[str] = None
    fournisseur: Optional[str] = None
    quantite: Optional[float] = Field(None, gt=0)
    prix_unitaire: Optional[float] = Field(None, gt=0)
    animal_id: Optional[int] = None
    lot_entomo_id: Optional[int] = None
    piece_jointe_url: Optional[str] = None


class DepenseCreate(DepenseBase):
    pass


class DepenseUpdate(BaseModel):
    categorie: Optional[CategorieDepenseEnum] = None
    montant: Optional[float] = Field(None, gt=0)
    date: Optional[date] = None # type: ignore
    description: Optional[str] = None
    fournisseur: Optional[str] = None
    quantite: Optional[float] = Field(None, gt=0)
    prix_unitaire: Optional[float] = Field(None, gt=0)


class DepenseResponse(DepenseBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class RecetteBase(BaseModel):
    categorie: CategorieRecetteEnum
    montant: float = Field(..., gt=0)
    date: date
    description: Optional[str] = None
    client: Optional[str] = None
    quantite: Optional[float] = Field(None, gt=0)
    prix_unitaire: Optional[float] = Field(None, gt=0)
    animal_id: Optional[int] = None
    lot_entomo_id: Optional[int] = None


class RecetteCreate(RecetteBase):
    pass


class RecetteUpdate(BaseModel):
    categorie: Optional[CategorieRecetteEnum] = None
    montant: Optional[float] = Field(None, gt=0)
    date: Optional[date] = None # type: ignore
    description: Optional[str] = None
    client: Optional[str] = None
    quantite: Optional[float] = Field(None, gt=0)
    prix_unitaire: Optional[float] = Field(None, gt=0)


class RecetteResponse(RecetteBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class AccountSummary(BaseModel):
    total_depenses: float
    total_recettes: float
    benefice: float
    marge_brute_pourcent: float
    depenses_par_categorie: dict
    recettes_par_categorie: dict
    tresorerie_previsionnelle: dict

File: frontend/src/lib/types/accounting.ts
// lib/types/accounting.ts
export type CategorieDepense =
    | 'achat_animaux' | 'achat_oeufs' | 'alimentation'
    | 'vaccins_soins' | 'equipement' | 'personnel'
    | 'eau_electricite' | 'entretien' | 'compostage'
    | 'transport' | 'frais_divers';

export type CategorieRecette =
    | 'vente_animaux_vivants' | 'vente_viande' | 'vente_lait'
    | 'vente_laine' | 'vente_oeufs' | 'vente_larves'
    | 'vente_compost' | 'vente_fumier' | 'subventions' | 'autres';

export interface DepenseBase {
    categorie: CategorieDepense;
    montant: number;
    date: string;
    description?: string;
    fournisseur?: string;
    quantite?: number;
    prix_unitaire?: number;
    animal_id?: number;
    lot_entomo_id?: number;
    piece_jointe_url?: string;
}

export interface DepenseCreate extends DepenseBase { }
export interface DepenseUpdate extends Partial<DepenseBase> { }

export interface DepenseResponse extends DepenseBase {
    id: number;
    created_at: string;
    updated_at: string;
}

export interface RecetteBase {
    categorie: CategorieRecette;
    montant: number;
    date: string;
    description?: string;
    client?: string;
    quantite?: number;
    prix_unitaire?: number;
    animal_id?: number;
    lot_entomo_id?: number;
}

export interface RecetteCreate extends RecetteBase { }
export interface RecetteUpdate extends Partial<RecetteBase> { }

export interface RecetteResponse extends RecetteBase {
    id: number;
    created_at: string;
    updated_at: string;
}

export interface AccountSummary {
    total_depenses: number;
    total_recettes: number;
    benefice: number;
    marge_brute_pourcent: number;
    depenses_par_categorie: Record<string, number>;
    recettes_par_categorie: Record<string, number>;
    tresorerie_previsionnelle: Record<string, number>;
}